### PR TITLE
Feat/rules

### PR DESCRIPTION
**Pull request by rudolfsonjunior**
Added Rules:

```json
"rules": {
  "max-char-commit-message": true,
  "max-char": 72,
  "min-char": 10,
  "end-with-dot": false
}
```

We have a file rulesConfig.js chcking these rules and return true or false.

in promptConfig.js attached to validate.

### CODE REVIEW

File: lib/rulesConfig.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import getConfig from './getConfig';
+
+const checkRulesMaxLength = (input) => {
+  const configuration = getConfig();
```
Comment (JPeer264) [inline lib/rulesConfig.js]: How about initialize `configuration` on top of the file?

File: lib/rulesConfig.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  if (configuration.rules['max-char-commit-message'] === true && input.length > configuration.rules['max-char']) {
+    return false;
+  } else if ((configuration.rules['max-char-commit-message'] === true && input.length <= configuration.rules['max-char']) || configuration.rules['max-char-commit-message'] === false) {
```
Comment (JPeer264) [inline lib/rulesConfig.js]: We never need a `else if`, one `if` is enough:

```js
if ((configuration.rules['max-char-commit-message'] === true && input.length <= configuration.rules['max-char']) || configuration.rules['max-char-commit-message'] === false) {
  return true;
}

return false;
```

File: lib/promptConfig.js [hunk 3]
```diff
@@ -29,7 +30,12 @@ const questions = (choicesList) => {
       type: 'input',
       name: 'description',
       message: 'Enter your commit message:',
-      validate: input => (input ? true : 'A commit message is mandatory!'),
+      validate: (input) => {
+        if (checkRulesEndWithDot(input) === false) return 'The commit message can not end with a dot';
+        else if (checkRulesMaxLength(input) === false) return `The commit message is not allowed to be longer as ${configuration.rules['max-char']}. Consider writing a body.`;
+        else if (checkRulesMinLength(input) === false) return `The commit message hast to be at least ${configuration.rules['min-char']} character.`;
+        return true;
```
Comment (JPeer264) [inline lib/promptConfig.js]: I think we should make this more flexible. If we add a rule we would probably not remember where to add some code. 

Maybe we can write it like following (I am open for better suggestions):

```js
const warnings = ruleWarningMessages(input);

if (warnings) return warnings;

return true;
```

With these functions we would never add any code again here, plus we can show more/all warnings at once. The current solution would always show one, although all of them would appear.

The rules could get configured like following:

```js
// lib/rules/availableRules.js
const rules = {
  endWithDot: 'The commit message...',
};
```

Other files:
`lib/rules/ruleWarningMessages.js`: in here we iterate over all available rules and check if the rule is set. The return would be a string with all messages in it (separated by `\n`). So if a new rule get added we never have to touch any function except `availableRules.js`.


What you think about it?

File: lib/promptConfig.js [hunk 4]
```diff
@@ -29,7 +30,12 @@ const questions = (choicesList) => {
       type: 'input',
       name: 'description',
       message: 'Enter your commit message:',
-      validate: input => (input ? true : 'A commit message is mandatory!'),
+      validate: (input) => {
+        if (checkRulesEndWithDot(input) === false) return 'The commit message can not end with a dot';
+        else if (checkRulesMaxLength(input) === false) return `The commit message is not allowed to be longer as ${configuration.rules['max-char']}. Consider writing a body.`;
+        else if (checkRulesMinLength(input) === false) return `The commit message hast to be at least ${configuration.rules['min-char']} character.`;
```
Comment (JPeer264) [inline lib/promptConfig.js]: spelling:
`commit message hast`-> `commit message  has`

**Comment by JPeer264** [comment]
Maybe we do not need `"max-char-commit-message": true,`. If either 'max-char' or 'min-char' is set, it is automatically true.

File: lib/promptConfig.js (hunk 3 above)
Comment (rudolfsonjunior) [inline lib/promptConfig.js]: That sounds good to me. I will make the changes as soon as possible.

**Comment by rudolfsonjunior** [comment]
Yes that is true, if we set a minimum or a maximum we do not really need a `max-char-commit-message`

File: lib/promptConfig.js (hunk 3 above)
Comment (JPeer264) [inline lib/promptConfig.js]: I just though about one small thing I forgot. Maybe the `availableRules.js` should look like following:

```js
const rules = {
  endWithDot: {
    message: 'The commit message...',
    isRule: (input) => {
      // some if statements
    }
};
```

Maybe `isRule` is not the best. But we need something when the rule should be active.

Comment (rudolfsonjunior) [inline lib/promptConfig.js]: Maybe we should name it like `passedCheck` or `passedRuleCheck`? So it looks like `rules.endWithDot.passedCheck(input)`  and this is either `true` or `false`?

Comment (JPeer264) [inline lib/promptConfig.js]: Maybe just `check`? 

Yes kind of : `rules.endWithDot` would be `rule` in a `forEach`

Comment (rudolfsonjunior) [inline lib/promptConfig.js]: I called it check for now.

File: lib/rules/ruleWarningMessages.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+const ruleWarningMessages = (input) => {
+  let warningMessage = '';
+  const configRuleEntries = entries(configuration.rules);
+
+  configRuleEntries.forEach((rule) => {
+    const camelCaseRuleName = (rule[0]).replace(/-([a-z])/g, g => (g[1].toUpperCase()));
+    const ruleIs = rules[camelCaseRuleName].check(input);
+
+    if (!ruleIs) {
```
Comment (JPeer264) [inline lib/rules/ruleWarningMessages.js]: Either we stick with `{ }` in one liner `if` or without as in `lib/rules/availableConfig` without `{ }`

File: lib/rules/ruleWarningMessages.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import entries from 'object.entries';
```
Comment (JPeer264) [inline lib/rules/ruleWarningMessages.js]: doesn't `Object.entries` work? I guess for that we do not need a external library

Comment (rudolfsonjunior) [inline lib/rules/ruleWarningMessages.js]: nope Object.entries does not work.

File: lib/rules/ruleWarningMessages.js (hunk 5 above)
Comment (rudolfsonjunior) [inline lib/rules/ruleWarningMessages.js]: I will change that

File: lib/rules/ruleWarningMessages.js (hunk 6 above)
Comment (rudolfsonjunior) [inline lib/rules/ruleWarningMessages.js]: TypeError: Object.entries is not a function

Comment (JPeer264) [inline lib/rules/ruleWarningMessages.js]: i tried it:

```js
import entries from 'object.entries';

entries(configuration.rules);
```

returned the same as:

```js
Object.entries(configuration.rules);
```

Comment (rudolfsonjunior) [inline lib/rules/ruleWarningMessages.js]: No, not really...
`W
  var configRuleEntries = Object.entries(configuration.rules);
                                 ^

TypeError: Object.entries is not a function`

Comment (JPeer264) [inline lib/rules/ruleWarningMessages.js]: which version of Node do you use?

Comment (rudolfsonjunior) [inline lib/rules/ruleWarningMessages.js]: where 

```js
import entries from 'object.entries';

entries(configuration.rules);
```

returns 

```js
[ [ 'max-char', 72 ],
  [ 'min-char', 10 ],
  [ 'end-with-dot', false ] ]
```

Comment (rudolfsonjunior) [inline lib/rules/ruleWarningMessages.js]: I use node version 6.8.1. But if this would be a problem, then we will never pass the node -v 4 travis test.

Comment (JPeer264) [inline lib/rules/ruleWarningMessages.js]: Well the tests passed in my case. Let's stay with the shim.

**Pull request merged**